File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/ml/processing/packer/TensorFromAudioPacker.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.ml.processing.packer;

import com.facebook.react.bridge.ReadableMap;
import java.nio.FloatBuffer;
import org.json.JSONException;
import org.json.JSONObject;
import org.pytorch.IValue;
import org.pytorch.Tensor;
import org.pytorch.rn.core.audio.IAudio;
import org.pytorch.rn.core.javascript.JSContext;
import org.pytorch.rn.core.ml.processing.Packer;
import org.pytorch.rn.core.ml.processing.PackerContext;

public class TensorFromAudioPacker implements Packer {

  private static final String AUDIO_KEY = "audio";

  @Override
  public IValue pack(JSONObject jobject, ReadableMap params, PackerContext packerContext)
      throws JSONException {
    final String key = jobject.getString(AUDIO_KEY);
    final ReadableMap jsRef = params.getMap(key);
    final IAudio audio = JSContext.unwrapObject(jsRef);

    final short[] audioData = audio.getData();
    final FloatBuffer floatBuffer = Tensor.allocateFloatBuffer(audioData.length);
    for (short audioDatum : audioData) {
      floatBuffer.put(audioDatum / (float) Short.MAX_VALUE);
    }

    final Tensor tensor = Tensor.fromBlob(floatBuffer, new long[] {1, audioData.length});
    return IValue.from(tensor);
  }
}

File: react-native-pytorch-core/ios/ML/Processing/Packer/TensorFromAudioPacker.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import SwiftyJSON

class TensorFromAudioPacker: Packer {
    func pack(modelSpec: JSON, params: NSDictionary, packerContext: PackerContext) throws -> IValue? {
        do {
            if let audioId = (params["audio"] as? NSDictionary)?["ID"] as? String {
                let audio = try JSContextUtils.unwrapObject(audioId, IAudio.self)
                let intArray = audio.getData().withUnsafeBytes { (pointer: UnsafeRawBufferPointer) -> [Int16] in
                    let buffer = UnsafeBufferPointer<Int16>(start: UnsafeRawPointer(pointer.baseAddress!)
                                                                .bindMemory(to: Int16.self, capacity: 1),
                                                     count: audio.getData().count / 2)
                    return [Int16](buffer)
                }
                let MAXVALUE = 32767
                var floatArray: [Float] = []
                for dat in intArray {
                    floatArray.append(Float(dat) / Float(MAXVALUE))
                }

                guard let tensor = Tensor.fromBlob(data: &floatArray,
                                                   shape: [1, NSNumber(value: floatArray.count)],
                                                   dtype: .float) else {
                    throw BaseIValuePackerError.audioUnwrapError
                }
                return IValue.fromTensor(tensor)
            } else {
                throw BaseIValuePackerError.audioUnwrapError
            }
        } catch {
            throw BaseIValuePackerError.audioUnwrapError
        }
    }
}
